arangod/Network/NetworkFeature.cpp

```diff
@@ -609,54 +816,13 @@ void NetworkFeature::finishRequest(network::ConnectionPool const& pool,
 void NetworkFeature::retryRequest(
     std::shared_ptr<network::RetryableRequest> req, RequestLane lane,
     std::chrono::steady_clock::duration duration) {
-  auto cb = [this, weak = std::weak_ptr(req)](bool cancelled) {
-    std::unique_lock guard(_workItemMutex);
-    if (auto self = weak.lock(); self) {
-      _retryRequests.erase(self);
-      guard.unlock();  // resuming the request does not access _retryRequests
-      if (cancelled) {
-        self->cancel();
-      } else {
-        self->retry();
-      }
-    }
-  };
-
-  // we need the mutex during `queueDelayed` because the lambda might be
-  // executed faster than we can add the work item to the _retryRequests map.
-  std::unique_lock guard(_workItemMutex);
-
-  // this will automatically cancel the request when we leave this
-  // method, unless we are canceling this scopeGuard explicitly.
-  auto cancelGuard = scopeGuard([req]() noexcept {
-    if (req) {
-      req->cancel();
-    }
-  });
-
-  if (server().isStopping()) {
-    // with trigger cancelGuard - cancels request
+  if (!req) {
     return;
   }
 
-  auto item = SchedulerFeature::SCHEDULER->queueDelayed(
-      "retry-requests", lane, duration, std::move(cb));
-
-  if (item != nullptr) {
-    try {
-      TRI_IF_FAILURE("NetworkFeature::retryRequestFail") {
-        THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
-      }
-
-      _retryRequests.emplace(std::move(req), std::move(item));
-      // the request successfully made it into _retryRequests.
-      // now abandon the automatic cancelation.
-      cancelGuard.cancel();
-    } catch (...) {
-    }
-  }
-  // if we get here and haven't canceled cancelGuard,
-  // the request will be automatically canceled
+  TRI_ASSERT(_retryThread);
```

Comment:
What if a retryable response comes back after the `NetworkFeature` has destroyed the retry thread? Could this then be a shutdown crash if we try to retry here? Should we check `_retryThread` here again (and maybe even use an atomic to check?)?

Reply:
Slightly adjusted it.